vpn fails to route packets to services on cluster

Hello dear friends.

I'm using kubevpn connecting on multiple k8s clusters.
I found it super convinent. 
Thank you for this awesome tool.


Recently after updating kubevpn to 2.7.8, it fails to route packets to services cluster.
Here are logs on vpn containers when it is failing


```
2025-05-12 02:22:42.477 gvisortunendpoint.go:142 info: [TUN-GVISOR] Add new route map TCP: 198.19.0.101 -> 127.0.0.1:10801-127.0.0.1:40490                                     
2025-05-12 02:22:42.477 gvisortunendpoint.go:104 Info: [gVISOR] recv udp 198.19.0.101:49463 -> 198.18.1.50:53 len:49 id:664c xsum: 0xda3f
2025-05-12 02:22:42.477 sniffer.go:145 Info: [gVISOR] recv udp 198.19.0.101:49463 -> 198.18.1.50:53 len:49 id:664c xsum: 0xda3f                                                 
2025-05-12 02:22:42.477 gvisorudpforwarder.go:23 info: [TUN-UDP] LocalPort: 53, LocalAddress: 198.18.1.50, RemotePort: 49463, RemoteAddress 198.19.0.101
2025-05-12 02:22:42.479 sniffer.go:145 Info: [gVISOR] send udp 198.18.1.50:53 -> 198.19.0.101:49463 len:96 id:707a xsum: 0xf07c                                                 
2025-05-12 02:22:42.479 gvisortunendpoint.go:27 Info: [gVISOR] send udp 198.18.1.50:53 -> 198.19.0.101:49463 len:96 id:707a xsum: 0xf07c
2025-05-12 02:22:42.485 gvisorudphandler.go:32 info: [TUN-UDP] LocalPort: 53, LocalAddress: 127.0.0.1, RemotePort: 0, RemoteAddress: 198.19.0.101                               
2025-05-12 02:22:48.829 gvisortunendpoint.go:128 warning: [TCP-GVISOR] No route for src: 198.19.0.101 -> dst: 198.19.151.170, drop it
2025-05-12 02:22:49.831 gvisortunendpoint.go:128 warning: [TCP-GVISOR] No route for src: 198.19.0.101 -> dst: 198.19.151.170, drop it                                           
2025-05-12 02:22:50.832 gvisortunendpoint.go:128 warning: [TCP-GVISOR] No route for src: 198.19.0.101 -> dst: 198.19.151.170, drop it
2025-05-12 02:22:51.832 gvisortunendpoint.go:128 warning: [TCP-GVISOR] No route for src: 198.19.0.101 -> dst: 198.19.151.170, drop it                                           
2025-05-12 02:22:52.477 gvisortunendpoint.go:104 Info: [gVISOR] recv udp 198.19.0.101:61044 -> 198.18.1.50:53 len:49 id:6bc4 xsum: 0x785c

```

I seems the connection between local pc(198.19.0.101) to vpn server has been established well, but it can not find Connection to destination service ip (198.19.151.170)

Can you give me any advice with this issue?




and as I was reading your code, I became curious about something.


https://github.com/kubenetworks/kubevpn/blob/master/pkg/core/gvisortunendpoint.go#L89-L92
`
h.addToRouteMapTCP(ctx, src, conn)
// inner ip like 198.19.0.100/102/103 connect each other
if config.CIDR.Contains(dst) || config.CIDR6.Contains(dst) {
	err = h.handlePacket(ctx, buf, read, src, dst, layers.IPProtocol(ipProtocol).String())
`
at these codes here. it stores connection first and handle packets
It stores with source ip,
https://github.com/kubenetworks/kubevpn/blob/master/pkg/core/gvisortunendpoint.go#L135
but it looks up with dst ip
https://github.com/kubenetworks/kubevpn/blob/master/pkg/core/gvisortunendpoint.go#L112

Maybe the issue has something to do with this part.

--- a/pkg/core/gvisortunendpoint.go
+++ b/pkg/core/gvisortunendpoint.go
@@ -88,7 +88,11 @@ func (h *gvisorTCPHandler) readFromTCPConnWriteToEndpoint(ctx context.Context, c
 
 		h.addToRouteMapTCP(ctx, src, conn)
 		// inner ip like 198.19.0.100/102/103 connect each other
-		if config.CIDR.Contains(dst) || config.CIDR6.Contains(dst) {
+		// for issue 594, sometimes k8s service network CIDR also use CIDR 198.19.151.170
+		// if we can find dst in route map, just trade packet as inner communicate
+		// if not find dst in route map, just trade packet as k8s service/pod ip
+		_, found := h.routeMapTCP.Load(dst.String())
+		if found && (config.CIDR.Contains(dst) || config.CIDR6.Contains(dst)) {
 			err = h.handlePacket(ctx, buf, read, src, dst, layers.IPProtocol(ipProtocol).String())
 			if err != nil {
 				plog.G(ctx).Errorf("[TCP-GVISOR] Failed to handle packet: %v", err)

